tachyon/crypto/hashes/sponge/sponge.h

```diff
@@ -0,0 +1,171 @@
+  static FieldElementSize Truncated(size_t num_bits) {
+    return {true, num_bits};
+  }
+
+  template <typename PrimeFieldTy>
+  size_t NumBits() {
+    static_assert(math::PrimeFieldTraits<PrimeFieldTy>::kIsPrimeField,
+                  "NumBits() is only supported for PrimeField");
+    if (is_truncated_) {
+      CHECK_LE(num_bits_, PrimeFieldTy::kModulusBits)
+          << "num_bits is greater than the capacity of the field.";
+    }
+    return PrimeFieldTy::kModulusBits - 1;
+  }
+
+  // Calculate the sum of prime field element sizes in |elements|.
+  template <typename PrimeFieldTy>
+  static size_t Sum(const std::vector<FieldElementSize>& elements) {
+    static_assert(math::PrimeFieldTraits<PrimeFieldTy>::kIsPrimeField,
+                  "Sum() is only supported for PrimeField");
+    return (PrimeFieldTy::kModulusBits - 1) * elements.size();
+  }
+
+  bool IsFull() const { return !is_truncated_; }
+  bool IsTruncated() const { return is_truncated_; }
+
+ private:
+  FieldElementSize(bool is_truncated, size_t num_bits = 0)
+      : is_truncated_(is_truncated), num_bits_(num_bits) {}
+
+  // If |is_truncated_| is false, sample field elements from the entire field.
+  // If |is_truncated_| is true, sample field elements from a subset of the
+  // field, specified by the maximum number of bit.
+  bool is_truncated_;
+  size_t num_bits_ = 0;
+};
+
+template <typename T>
+struct CryptographicSpongeTraits;
+
+// The interface for a cryptographic sponge.
+// A sponge can |Absorb| and later |Squeeze| bytes of field elements.
+// The outputs are dependent on previous |Absorb| and |Squeeze| calls.
+template <typename Derived>
+class CryptographicSponge {
+ public:
+  using F = typename CryptographicSpongeTraits<Derived>::F;
+
+  // Squeeze |num_elements| nonnative field elements from the sponge.
+  std::vector<F> SqueezeFieldElements(size_t num_elements) {
+    Derived* derived = static_cast<Derived*>(this);
+    return derived->SqueezeFieldElementsWithSizes(base::CreateVector(
+        num_elements, []() { return FieldElementSize::Full(); }));
+  }
+
+  // Creates a new sponge with applied domain separation.
+  Derived Fork(const absl::Span<const uint8_t>& domain) const {
+    const Derived* derived = static_cast<const Derived*>(this);
+    CHECK(derived->Absorb(domain));
+    return *derived;
+  }
+
+ protected:
+  std::vector<F> SqueezeFieldElementsWithSizesDefaultImpl(
+      const std::vector<FieldElementSize>& sizes) {
+    if constexpr (math::PrimeFieldTraits<F>::kIsPrimeField) {
+      if (sizes.empty()) {
+        return {};
+      }
+
+      size_t total_num_bits = FieldElementSize::Sum<F>(sizes);
+
+      Derived derived = static_cast<Derived*>(this);
+      std::vector<bool> bits = derived->SqueezeBits(total_num_bits);
+      auto bits_window = bits.begin();
+
+      std::vector<F> output;
+      output.reserve(sizes.size());
+      for (const FieldElementSize& size : sizes) {
+        size_t num_bits = F::kModulusBits - 1;
```

Comment:
I will fix this line because of the following compile error in MacOS:
`error: unused variable 'size' [-Werror,-Wunused-variable]`
It's better to use the `NumBits()` to check the constraints of `FieldElementSize` regardless of the error though.
```suggestion
      for (const FieldElementSize& size : sizes) {
        size_t num_bits = size.NumBits<F>();
```